src/Nethermind/Nethermind.Evm/Tracing/Debugger/DebugTracer.cs

```diff
@@ -0,0 +1,252 @@
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Evm.Tracing.GethStyle;
+using Nethermind.Int256;
+using Nethermind.State;
+
+namespace Nethermind.Evm.Tracing.DebugTrace;
+internal class DebugTracer : ITxTracer
+{
+    public DebugTracer(ITxTracer tracer)
+    {
+        InnerTracer = tracer;
+    }
+
+    public ITxTracer InnerTracer { get; private set; }
+    private ManualResetEventSlim _manualResetEvent = new ManualResetEventSlim(true, 0);
+
+    public bool IsTracingReceipt => ((ITxTracer)InnerTracer).IsTracingReceipt;
+
+    public bool IsTracingActions => ((ITxTracer)InnerTracer).IsTracingActions;
+
+    public bool IsTracingOpLevelStorage => ((ITxTracer)InnerTracer).IsTracingOpLevelStorage;
+
+    public bool IsTracingMemory => ((ITxTracer)InnerTracer).IsTracingMemory;
+
+    public bool IsTracingInstructions => ((ITxTracer)InnerTracer).IsTracingInstructions;
+
+    public bool IsTracingRefunds => ((ITxTracer)InnerTracer).IsTracingRefunds;
+
+    public bool IsTracingCode => ((ITxTracer)InnerTracer).IsTracingCode;
+
+    public bool IsTracingStack => ((ITxTracer)InnerTracer).IsTracingStack;
+
+    public bool IsTracingBlockHash => ((ITxTracer)InnerTracer).IsTracingBlockHash;
+
+    public bool IsTracingAccess => ((ITxTracer)InnerTracer).IsTracingAccess;
+
+    public bool IsTracingFees => ((ITxTracer)InnerTracer).IsTracingFees;
+
+    public bool IsTracingState => ((IStateTracer)InnerTracer).IsTracingState;
+
+    public bool IsTracingStorage => ((IStorageTracer)InnerTracer).IsTracingStorage;
+
+    private object _lock = new();
+    private Dictionary<int, Func<EvmState, bool>> _breakPoints = new();
+    public bool IsStepByStepModeOn { get; set; } = false;
+    public EvmState CurrentState;
+    public bool TryWait(EvmState evmState)
+    {
+        lock (_lock)
+        {
+            CurrentState = evmState;
+        }
+
+        if (IsStepByStepModeOn)
+        {
+            _manualResetEvent.Reset();
+            _manualResetEvent.Wait();
+            return true;
+        }
+        else
+        {
+            return CheckBreakPoint();
+        }
+
+    }
+
+    public void MoveNext() => _manualResetEvent.Set();
+    public bool CanReadState() => !_manualResetEvent.IsSet;
+    public void SetBreakPoint(int programCounter, Func<EvmState, bool> condition = null)
+        => _breakPoints.Add(programCounter, condition);
+    public bool CheckBreakPoint()
+    {
+        if (_breakPoints.ContainsKey(CurrentState.ProgramCounter))
+        {
+            Func<EvmState, bool> condition = _breakPoints[CurrentState.ProgramCounter];
+            bool conditionResults = condition is null ? true : condition.Invoke(CurrentState);
+            if (conditionResults)
+            {
+                _manualResetEvent.Reset();
+                _manualResetEvent.Wait();
```

Comment:
So you want more AutoResetEvent? Though there is no Slim. You could use SemaphoreSlim, or just Monitor

Reply:
Was thinking the same thing